### PR TITLE
Wist core webassembly based parser

### PR DESCRIPTION
**Pull request by nishtahir**
Port the base parser to C++ and WebAssembly. Introduce `wist-fmt` code formatter. Addresses #71 and #83

### CODE REVIEW

**Comment by nishtahir** [comment]
This is aimed at fixing #71. All of the existing rules have been ported over to use the new light AST. 

Here are some minor Benchmarks that I'm currently seeing:

MacBook Pro (Retina, 15-inch, Mid 2015)
Test file: `test.brs` 1048 loc

```
$ node -v
v10.2.1

$ node bin/wist.js test.brs
✖ 169 problems (12 errors, 157 warnings)

Finished in 6.89s

$ wist test.brs
✖ 164 problems (12 errors, 152 warnings)

Finished in 17.18s
```

The expectation is that a single brightscript file will not have nearly that much code. A more realistic situation with [this](https://raw.githubusercontent.com/BrightcoveOS/Roku-Sample-App/master/source/BrightcoveMediaAPI.brs) ~150 loc file.

```
$ node bin/wist.js test.brs
✖ 38 problems (3 errors, 35 warnings)

Finished in 2.30s

$ wist test.brs
✖ 33 problems (3 errors, 30 warnings)

Finished in 7.73s
```

I think where the next step would be trying to optimize the Javascript side of the tool

File: package.json [hunk 1]
```diff
@@ -55,14 +52,9 @@
...
-    "pre-commit": "1.2.2"
-  },
-  "ava": {
-    "files": [
-      "test/**/*.js"
-    ],
-    "source": [
-      "lib/**/*.js"
-    ]
+    "ts-loader": "^4.2.0",
```
Comment (vannuysm) [inline package.json]: Are we actually using these 4 dependencies? I thought we took out webpack and typescript.

File: src/js/cli-engine.js [hunk 2]
```diff
@@ -139,14 +139,21 @@ class CLIEngine {
             return null;
         }
 
-        const formatterPath = `./formatters/${format}`;
```
Comment (vannuysm) [inline src/js/cli-engine.js]: What is the purpose of this change?

File: package.json (hunk 1 above)
Comment (nishtahir) [reply package.json]: Good catch 👍

File: src/js/cli-engine.js (hunk 2 above)
Comment (nishtahir) [reply src/js/cli-engine.js]: This happened around the period we were experimenting with Webpack. I ended up hardcoding the paths in the require statement to get it to load properly. I will revert.

File: src/js/rules/camel-case-func.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+const shims = require('../util/shims');
+
+module.exports = {
+    create(context) {
+        const report = (astNode) => {
+            astNode.children.forEach(element => {
```
Comment (vannuysm) [inline src/js/rules/camel-case-func.js]: Instead of iterating over every child of the AST here, couldn't we just look at the node at position 1 like we were before? If not, could we at least exit early so that we don't continue iterating?

Comment (nishtahir) [reply src/js/rules/camel-case-func.js]: Hmm this was added before I implemented some of the newer shims. Updating to use `findFirst` which exits the loop early.

**Comment by cdthompson** [comment]
nish, I tried to build the package and install it into my node_modules but its missing the /dist directory and contents when i do it that way.  Then wist fails to run, showing the following error:
```
Error: ENOENT: no such file or directory, open 'dist/libwist.wasm'
```

Any ideas?

**Comment by vannuysm** [comment]
This LGTM now. Do you think we can squash some of the commits (especially the WIP ones) into a few easy to distinguish units before merging?

**Comment by nishtahir** [comment]
Yeah good plan, I already had plans to do this 👍
